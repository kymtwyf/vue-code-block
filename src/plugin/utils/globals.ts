export const pluginName = 'v-code-block';
