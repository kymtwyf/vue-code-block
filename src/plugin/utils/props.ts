
export const AllProps = {
	browserWindow: false,
	code: '',
	codeBlockRadius: '0.5rem',
	copyButton: true,
	copyIcons: true,
	copyTab: true,
	copyFailedText: 'Copy failed!',
	copyText: 'Copy Code',
	copySuccessText: 'Copied!',
	floatingTabs: true,
	height: 'auto',
	highlightjs: false,
	indent: 2,
	label: '',
	lang: 'javascript',
	maxHeight: 'auto',
	persistentCopyButton: false,
	prismjs: false,
	prismPlugin: false,
	runTab: false,
	runText: 'Run',
	tabGap: '0.25rem',
	tabs: false,
	theme: 'neon-bunny',
};
