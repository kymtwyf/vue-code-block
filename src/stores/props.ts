import { defineStore } from 'pinia';
// @ts-expect-error @ts-ignore
import packageJson from '@root/package.json';

const dependencies = packageJson.dependencies;
const prismVersion = dependencies.prismjs.replace('^', '');
const highlightJsVersion = dependencies['highlight.js'].replace('^', '');

const highlightJsLinks = {
	cdn: `https://cdn.jsdelivr.net/gh/highlightjs/cdn-release@${highlightJsVersion}/build/styles/`,
	homepage: 'https://highlightjs.org/',
	themes: `https://github.com/highlightjs/highlight.js/tree/${highlightJsVersion}/src/styles`,
};
const prismLinks = {
	cdn: `https://cdn.jsdelivr.net/gh/PrismJS/prism@${prismVersion}/themes/`,
	homepage: 'https://prismjs.com/',
	prismThemes: `https://github.com/PrismJS/prism/tree/v${prismVersion}/themes`,
	themes: 'https://github.com/PrismJS/prism-themes',
};


const propsHeaders = [
	{
		align: 'start',
		filterable: true,
		key: 'name',
		sortable: true,
		title: 'Name',
		width: '20%',
	},
	{
		align: 'start',
		filterable: false,
		key: 'type',
		sortable: false,
		title: 'Type',
		width: '15%',
	},
	{
		align: 'start',
		filterable: false,
		key: 'default',
		sortable: false,
		title: 'Default',
		width: '10%',
	},
	{
		align: 'start',
		filterable: false,
		key: 'desc',
		sortable: false,
		title: 'Description',
	},
];

const componentProps = [
	{
		default: 'false',
		desc: 'To give the code block a browser window look.',
		name: 'browserWindow',
		required: false,
		type: 'boolean',
	},
	{
		default: 'undefined',
		desc: 'Used to load the css from a specified location instead of the component using the CDN location. If you want to dynamically change the theme, you will also need to change the theme prop.',
		name: 'cssPath',
		required: false,
		type: 'string | undefined',
	},
	{
		default: '',
		desc: 'The code to be displayed in the code block.',
		name: 'code',
		required: true,
		type: 'object | string | number',
	},
	{
		default: '0.5rem',
		desc: 'The border radius of the code block.',
		name: 'codeBlockRadius',
		required: false,
		type: 'string',
	},
	{
		default: 'true',
		desc: 'To show a copy button within the code block.',
		name: 'copyButton',
		required: false,
		type: 'boolean',
	},
	{
		default: 'true',
		desc: 'To show the copy icons on the copy code tab.',
		name: 'copyIcons',
		required: false,
		type: 'boolean',
	},
	{
		default: 'true?',
		desc: 'To show the copy code tab.',
		name: 'copyTab',
		required: false,
		type: 'boolean',
	},
	{
		default: 'Copy failed!',
		desc: 'The text to be displayed when the code copy failed.',
		name: 'copyFailedText',
		required: false,
		type: 'string',
	},
	{
		default: 'Copy Code',
		desc: 'The text to be displayed on the copy tab.',
		name: 'copyText',
		required: false,
		type: 'string',
	},
	{
		default: 'Copied!',
		desc: 'The text to be displayed when the code copy was successful.',
		name: 'copySuccessText',
		required: false,
		type: 'string',
	},
	{
		default: 'true',
		desc: 'To make the tabs float on top of the code block.',
		name: 'floatingTabs',
		required: false,
		type: 'boolean',
	},
	{
		default: 'auto',
		desc: 'The height of the code block.',
		name: 'height',
		required: false,
		type: 'string | number',
	},
	{
		default: 'false',
		desc: 'Specifies that the Highlight.js library should be used. * Required if <code class="ic">prismjs</code> prop is not set.',
		name: 'highlightjs',
		required: false,
		type: 'boolean',
	},
	{
		default: '4',
		desc: 'The number of spaces to indent the code for json.',
		name: 'indent',
		required: false,
		type: 'number',
	},
	{
		default: '',
		desc: 'The label to be displayed on the code block.',
		name: 'label',
		required: false,
		type: 'string',
	},
	{
		default: 'javascript',
		desc: 'The language of the code.',
		name: 'lang',
		required: false,
		type: 'string',
	},
	{
		default: 'auto',
		desc: 'The max height of the code block.',
		name: 'maxHeight',
		required: false,
		type: 'string | number',
	},
	{
		default: 'false',
		desc: 'To show a persistent copy button within the code block.',
		name: 'persistentCopyButton',
		required: false,
		type: 'boolean',
	},
	{
		default: 'true',
		desc: 'Specifies that the PrismJS library should be used. * Required if <code class="ic">highlightjs</code> prop is not set.',
		name: 'prismjs',
		required: false,
		type: 'boolean',
	},
	{
		default: 'false',
		desc: 'Specifies that a PrismJS plugin is being used. This is needed for the plugin to work properly.',
		name: 'prismPlugin',
		required: false,
		type: 'boolean',
	},
	{
		default: 'false',
		desc: 'To show the run tab.',
		name: 'runTab',
		required: false,
		type: 'boolean',
	},
	{
		default: 'Run',
		desc: 'The text to be displayed on the run tab.',
		name: 'runText',
		required: false,
		type: 'string',
	},
	{
		default: '0.25rem',
		desc: 'The gap between the tabs.',
		name: 'tabGap',
		required: false,
		type: 'string',
	},
	{
		default: 'false',
		desc: 'To show the tabs. This will show/hide all tabs.',
		name: 'tabs',
		required: false,
		type: 'boolean',
	},
	{
		default: 'neon-bunny',
		desc: `<p>The theme to be used for the code block. Available options include:</p>
		<p>
			<code>false</code> - When you are loading the theme's styles yourself.
		</p>
		<p>
			<strong>Neon Bunny Themes</strong>
			<br />
			<code>neon-bunny</code>
			<br />
			<code>neon-bunny-carrot</code>
		</p>
		<p>
		<a href="${prismLinks.prismThemes}" target="_blank">PrismJS Themes</a>
		</p>
		<p>
		  <a href="${highlightJsLinks.themes}" target="_blank">Highlight.js Themes</a>
		</p>
		`,
		name: 'theme',
		required: false,
		type: 'string | boolean',
	},
];

// eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars
const componentProps2 = {
	// theme: {
	// 	type: 'string | boolean',
	// 	required: false,
	// 	id: 'props-theme-option',
	// 	default: 'neon-bunny',
	// 	description: `<p>The theme to be used for the code block. Available options include:</p>
	// 	<p>
	// 		<code>false</code> - When you are loading the theme's styles yourself.
	// 	</p>
	// 	<p>
	// 		<strong>Neon Bunny Themes</strong>
	// 		<br />
	// 		<code>neon-bunny</code>
	// 		<br />
	// 		<code>neon-bunny-carrot</code>
	// 	</p>
	// 	<p>
	// 		<a href="${prismLinks.prismThemes}" target="_blank">PrismJS Themes</a>
	// 	</p>
	// 	<p>
	// 		<a href="${highlightJsLinks.themes}" target="_blank">Highlight.js Themes</a>
	// 	</p>
	// 	`,
	// }
};


export const usePropsStore = defineStore('props', {
	state: () => {
		return {
			componentProps,
			highlightJsLinks,
			prismLinks,
			propsHeaders,
		};
	},
});
