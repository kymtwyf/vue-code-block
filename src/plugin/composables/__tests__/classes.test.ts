import { describe, it, expect, test } from 'vitest';
import { ref } from 'vue';
import {
	useCodeBlockClasses,
	useCopyButtonClasses,
	useIconClasses,
	useLabelClasses,
	useTabClasses,
} from '../classes';
import {
	CopyStatus,
	Props,
	UseTheme,
} from '../../types';
import { AllProps } from '../../utils/props';
import { pluginName } from '../../utils/globals';


const defaultProps: Props = { ...AllProps };
const defaultTheme = defaultProps.theme as UseTheme;


describe('Classes Composable', () => {
	describe('useCodeBlockClasses', () => {
		const isMobile = ref(false);
		const isMobileClass = `${pluginName}--mobile`;
		const isPrism = ref(false);
		const isPrismClass = `${pluginName}--prismjs`;
		let classes = useCodeBlockClasses({ isMobile, isPrism });

		it('should return the codeblock class defaults', () => {
			expect(classes).toMatchSnapshot();
		});

		test('mobile classes', () => {
			isMobile.value = true;

			classes = useCodeBlockClasses({ isMobile, isPrism });
			expect(classes[isMobileClass]).toBeTruthy();

			isMobile.value = false;
			classes = useCodeBlockClasses({ isMobile, isPrism });
			expect(classes[isMobileClass]).toBeFalsy();
		});

		test('prismjs classes', () => {
			isPrism.value = true;

			classes = useCodeBlockClasses({ isMobile, isPrism });
			expect(classes[isPrismClass]).toBeTruthy();

			isPrism.value = false;
			classes = useCodeBlockClasses({ isMobile, isPrism });
			expect(classes[isPrismClass]).toBeFalsy();
		});
	});

	describe('useCopyButtonClasses', () => {
		const copyStatus = ref<CopyStatus>('copy');
		const copyStatusClassPrefix = `${pluginName}--code-copy-button-status`;
		const isMobile = ref(false);
		const isMobileClass = `${pluginName}--code-copy-button-mobile`;
		const persistentCopyButton = ref(defaultProps.persistentCopyButton);
		const persistentCopyButtonCLass = `${pluginName}--code-copy-button-persist`;

		let classes = useCopyButtonClasses({ copyStatus, isMobile, persistentCopyButton });

		it('should return copy button class defaults', () => {
			expect(classes).toMatchSnapshot();
		});

		test('copy status classes', () => {
			classes = useCopyButtonClasses({ copyStatus, isMobile, persistentCopyButton });
			expect(classes[`${copyStatusClassPrefix}-copy`]).toBeTruthy();

			copyStatus.value = 'success';
			classes = useCopyButtonClasses({ copyStatus, isMobile, persistentCopyButton });
			expect(classes[`${copyStatusClassPrefix}-success`]).toBeTruthy();

			copyStatus.value = 'failed';
			classes = useCopyButtonClasses({ copyStatus, isMobile, persistentCopyButton });
			expect(classes[`${copyStatusClassPrefix}-failed`]).toMatchInlineSnapshot(`true`);
		});

		test('mobile classes', () => {
			isMobile.value = true;
			classes = useCopyButtonClasses({ copyStatus, isMobile, persistentCopyButton });
			expect(classes[isMobileClass]).toBeTruthy();

			isMobile.value = false;
			classes = useCopyButtonClasses({ copyStatus, isMobile, persistentCopyButton });
			expect(classes[isMobileClass]).toBeFalsy();
		});

		test('persistent copy button classes', () => {
			persistentCopyButton.value = true;
			classes = useCopyButtonClasses({ copyStatus, isMobile, persistentCopyButton });
			expect(classes[persistentCopyButtonCLass]).toBeTruthy();

			persistentCopyButton.value = false;
			classes = useCopyButtonClasses({ copyStatus, isMobile, persistentCopyButton });
			expect(classes[persistentCopyButtonCLass]).toBeFalsy();
		});
	});

	describe('useIconClasses', () => {
		const copyStatus = ref<CopyStatus>('copy');
		const highlightjs = ref(defaultProps.highlightjs);
		const highlightjsStatusClassPrefix = `${pluginName}--tab-highlightjs-${defaultTheme}-icon-status`;
		const prismCopyStatusClassPrefix = `${pluginName}--tab-prism-${defaultTheme}-icon-status`;
		const prismjsClass = `${pluginName}--tab-prism-neon-bunny-icon`;
		const highlightjsClass = `${pluginName}--tab-highlightjs-neon-bunny-icon`;

		let classes = useIconClasses({ copyStatus, highlightjs, useTheme: defaultTheme });

		it('should return icon class defaults', () => {
			expect(classes).toMatchSnapshot();
		});

		test('highlightjs class', () => {
			highlightjs.value = true;
			classes = useIconClasses({ copyStatus, highlightjs, useTheme: defaultTheme });
			expect(classes[highlightjsClass]).toBeTruthy();

			highlightjs.value = false;
			classes = useIconClasses({ copyStatus, highlightjs, useTheme: defaultTheme });
			expect(classes[highlightjsClass]).toBeFalsy();
		});

		test('prismjs class', () => {
			classes = useIconClasses({ copyStatus, highlightjs, useTheme: defaultTheme });
			expect(classes[prismjsClass]).toBeTruthy();

			highlightjs.value = true;
			classes = useIconClasses({ copyStatus, highlightjs, useTheme: defaultTheme });
			expect(classes[prismjsClass]).toBeFalsy();
		});

		test('copy status classes for prismjs', () => {
			highlightjs.value = false;
			classes = useIconClasses({ copyStatus, highlightjs, useTheme: defaultTheme });
			expect(classes[`${prismCopyStatusClassPrefix}-copy`]).toBeTruthy();

			copyStatus.value = 'success';
			classes = useIconClasses({ copyStatus, highlightjs, useTheme: defaultTheme });
			expect(classes[`${prismCopyStatusClassPrefix}-success`]).toBeTruthy();

			copyStatus.value = 'failed';
			classes = useIconClasses({ copyStatus, highlightjs, useTheme: defaultTheme });
			expect(classes[`${prismCopyStatusClassPrefix}-failed`]).toBeTruthy();
		});

		test('copy status classes for highlightjs', () => {
			highlightjs.value = true;
			copyStatus.value = 'copy';
			classes = useIconClasses({ copyStatus, highlightjs, useTheme: defaultTheme });
			expect(classes[`${highlightjsStatusClassPrefix}-copy`]).toBeTruthy();

			copyStatus.value = 'success';
			classes = useIconClasses({ copyStatus, highlightjs, useTheme: defaultTheme });
			expect(classes[`${highlightjsStatusClassPrefix}-success`]).toBeTruthy();

			copyStatus.value = 'failed';
			classes = useIconClasses({ copyStatus, highlightjs, useTheme: defaultTheme });
			expect(classes[`${highlightjsStatusClassPrefix}-failed`]).toBeTruthy();
		});
	});

	describe('useLabelClasses', () => {
		const isMobile = ref<boolean>(false);
		const mobileClass = `${pluginName}--label-mobile`;

		let classes = useLabelClasses({ isMobile });

		it('should return icon class defaults', () => {
			expect(classes).toMatchSnapshot();
		});

		test('mobile classes', () => {
			classes = useLabelClasses({ isMobile });
			expect(classes[mobileClass]).toBeFalsy();

			isMobile.value = true;
			classes = useLabelClasses({ isMobile });
			expect(classes[mobileClass]).toBeTruthy();
		});
	});

	describe('useTabClasses', () => {
		const highlightjs = ref(defaultProps.highlightjs);
		const tabClassPrefix = `${pluginName}--tab`;

		let classes = useTabClasses({ highlightjs, useTheme: defaultTheme });

		it('should return icon class defaults', () => {
			expect(classes).toMatchSnapshot();
		});

		test('highlightjs classes', () => {
			classes = useTabClasses({ highlightjs, useTheme: defaultTheme });
			expect(classes[`${tabClassPrefix}-highlightjs-${defaultTheme}`]).toBeFalsy();

			highlightjs.value = true;
			classes = useTabClasses({ highlightjs, useTheme: defaultTheme });
			expect(classes[`${tabClassPrefix}-highlightjs-${defaultTheme}`]).toBeTruthy();
		});

		test('prismjs classes', () => {
			highlightjs.value = false;
			classes = useTabClasses({ highlightjs, useTheme: defaultTheme });
			expect(classes[`${tabClassPrefix}-prism-${defaultTheme}`]).toBeTruthy();

			highlightjs.value = true;
			classes = useTabClasses({ highlightjs, useTheme: defaultTheme });
			expect(classes[`${tabClassPrefix}-prism-${defaultTheme}`]).toBeFalsy();
		});
	});
});
