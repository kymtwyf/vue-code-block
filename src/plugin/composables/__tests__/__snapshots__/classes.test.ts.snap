// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Classes Composable > useCodeBlockClasses > should return the codeblock class defaults 1`] = `
{
  "v-code-block": true,
  "v-code-block--highlightjs": true,
  "v-code-block--mobile": false,
  "v-code-block--prismjs": false,
  "v-code-block-mb-5": true,
}
`;

exports[`Classes Composable > useCopyButtonClasses > should return copy button class defaults 1`] = `
{
  "v-code-block--code-copy-button": true,
  "v-code-block--code-copy-button-mobile": false,
  "v-code-block--code-copy-button-persist": false,
  "v-code-block--code-copy-button-status-copy": true,
}
`;

exports[`Classes Composable > useIconClasses > should return icon class defaults 1`] = `
{
  "v-code-block--button-copy-icon-status-copy": true,
  "v-code-block--me-1": true,
  "v-code-block--tab-prism-neon-bunny-icon": true,
  "v-code-block--tab-prism-neon-bunny-icon-status-copy": true,
}
`;

exports[`Classes Composable > useLabelClasses > should return icon class defaults 1`] = `
{
  "v-code-block--label": true,
  "v-code-block--label-mobile": false,
}
`;

exports[`Classes Composable > useTabClasses > should return icon class defaults 1`] = `
{
  "v-code-block--tab-neon-bunny": true,
  "v-code-block--tab-prism-neon-bunny": true,
}
`;
